### PR TITLE
Customize 3D toolbar

### PR DESCRIPTION
**Pull request by PennyQ**
This code is still under cleaning and scatter plot & isosurface part haven't finished yet. 

In common toolbar: 
-  `class VispyViewerToolbar(BasicToolBar)` is not inherited from previous `QtWidgets.QToolBar`
- three common tools - save, rotate and record - are also rewrote to `glue.viewers.common.qt.tool` style.
-  `class VispyMouseMode(CheckableTool)` is an abstract selection mode class to handle mouse events and draw selection shape

In Volume/Isosurface/Scatter viewer, the toolbar is similar to previous but now inherited from new common toolbar class, 3D selection modes are now specific `VispyMouseMode` instance.

### CODE REVIEW

File: glue_vispy_viewers/common/tests/test_vispy_toolbar.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    session = simple_session()
+    with warnings.catch_warnings(record=True) as w:
+        viewer = ExampleViewer(session)
+        toolbar = viewer.toolbar
+
+        # TODO: how to add a test for dialogue option?
+        # test save tool
+        # toolbar.actions['Save'].trigger()
+        # assert toolbar.active_tool.tool_id == 'Save'
+
```
Comment (PennyQ) [inline glue_vispy_viewers/common/tests/test_vispy_toolbar.py:47]: @astrofrog the `save` trigger will pop up a dialogue and requires to choose between save/cancel, but in either case I will get an error and the test process is stopped, do you know how to test it properly? Thanks!

**Comment by PennyQ** [comment]
@astrofrog finally done ;) can you have a look and let me know which part I could improve ? Thanks!

**Comment by astrofrog** [comment]
@PennyQ - sorry for the delay, can you rebase this and then try and make sure the tests pass?

**Comment by PennyQ** [comment]
@astrofrog Done ;)

**Comment by PennyQ** [comment]
@astrofrog There is a very strange bug that 'rectangle' and 'circular' selection works well before switching to 'lasso' tool, the selection line got drawn and ROI get highlighted. But after using 'lasso' tool, the selection line (of both rectangle and circular tool) never got drawn, while ROI region is still highlighted :(

I am sure the line_pos for drawing is valid and could be drawn in an individual vispy script and it's within the canvas size, also update() function of line visual didn't solve it. Could you take a look at it? Thanks!!!

**Comment by astrofrog** [comment]
@PennyQ - I'm taking a look at this now - thanks!

**Comment by astrofrog** [comment]
@PennyQ - I'm still working on this (as you can see, I've pushed some changes). I'm going to work on the volume selections next - I want to try and avoid duplicating code between scatter and volume viewers.

**Comment by astrofrog** [comment]
@PennyQ - I went a bit crazy with the refactoring :D Now the selection tools are the same for the volume and scatter plots, and as a bonus, if you have multiple datasets in a viewer, the selection works properly (instead of picking one dataset).

So this fixes https://github.com/glue-viz/glue-vispy-viewers/issues/220 and https://github.com/glue-viz/glue-vispy-viewers/issues/93

**Comment by astrofrog** [comment]
Remaining things to do:
- [x] Make sure we can save/load the `MultiElementSubsetState` objects (this should go in the core package)
- [x] Add changelog entry

I can work on this during travel this weekend, then we can merge this.

**Comment by astrofrog** [comment]
@PennyQ - I think this is now good to go so I will merge once tests pass. Thanks for your initial work on this! :)

File: glue_vispy_viewers/common/tests/test_vispy_toolbar.py [hunk 2]
```diff
@@ -33,16 +33,16 @@ def callback(self, mode):
...
         viewer = ExampleViewer(session)
         toolbar = viewer.toolbar
 
-        # TODO: how to add a test for dialogue option?
         # test save tool
-        # toolbar.actions['Save'].trigger()
-        # assert toolbar.active_tool.tool_id == 'Save'
+        with patch('qtpy.compat.getsavefilename') as fd:
+            fd.return_value = tmpdir.join('test.png').strpath, 'jnk'
+            toolbar.actions['vispy:save'].trigger()
```
Comment (astrofrog) [inline glue_vispy_viewers/common/tests/test_vispy_toolbar.py:46]: @PennyQ - just for info, this is how you can test things that use e.g. the save dialog. This replaces the normal save dialog function with a 'fake' one that just returns a filename.

**Comment by astrofrog** [comment]
Before I merge this, I need to fix the 'record' tool and if possible test it.

File: glue_vispy_viewers/common/tests/test_vispy_toolbar.py (hunk 2 above)
Comment (PennyQ) [reply glue_vispy_viewers/common/tests/test_vispy_toolbar.py:46]: @astrofrog this is awesome, thanks for letting me know!

**Pull request merged**